src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java

```diff
@@ -235,6 +284,81 @@ private int addProbabilisticSplits(
 		return i;
 	}
 
+	private List<InputSplit> filterByInterval(List<InputSplit> splits, Configuration conf)
+			throws IOException {
+		List<Locatable> intervals = getIntervals(conf);
+		if (intervals == null) {
+			return splits;
+		}
+
+		// Get the chunks in the intervals we want (chunks give start and end file pointers
+		// into a BAM file) by looking in all the indexes for the BAM fils
+		List<Chunk> chunks = new ArrayList<Chunk>();
+		Set<Path> bamFiles = new LinkedHashSet<Path>();
+		for (InputSplit split : splits) {
+			bamFiles.add(((FileVirtualSplit) split).getPath());
+		}
+		for (Path bamFile : bamFiles) {
+			Path indexFile = bamFile.suffix(BAMIndex.BAMIndexSuffix);
+			FileSystem fs = bamFile.getFileSystem(conf);
+			if (!fs.exists(indexFile)) {
+				System.err.println("WARNING: no BAM index file found, splits will not be " +
+						"filtered, which may be very inefficient: " + indexFile);
+				return splits;
+			}
+
+			FSDataInputStream in = fs.open(bamFile);
+			SAMFileHeader header = SAMHeaderReader.readSAMHeaderFrom(in, conf);
+			SAMSequenceDictionary dict = header.getSequenceDictionary();
+
+			WrapSeekable seekableIdxFile = WrapSeekable.openPath(conf, indexFile);
+			PublicDiskBasedBAMFileIndex idx = new PublicDiskBasedBAMFileIndex(seekableIdxFile,
+					dict);
+
+			for (Locatable interval : intervals) {
+				int referenceIndex = dict.getSequenceIndex(interval.getContig());
+				int intervalStart = interval.getStart();
+				int intervalEnd = interval.getEnd();
+				BAMFileSpan sfs = idx.getSpanOverlapping(referenceIndex, intervalStart, intervalEnd);
+				chunks.addAll(sfs.getChunks());
+			}
+		}
+
+		// Use the chunks to filter the splits
+		List<InputSplit> filteredSplits = new ArrayList<InputSplit>();
+		for (InputSplit split : splits) {
+			FileVirtualSplit virtualSplit = (FileVirtualSplit) split;
+			long splitStart = virtualSplit.getStartVirtualOffset();
+			long splitEnd = virtualSplit.getEndVirtualOffset();
+			// if any chunk overlaps with the split, keep the split, but adjust the start and
+			// end to the maximally overlapping portion for all chunks that overlap
+			long newStart = Long.MAX_VALUE;
+			long newEnd = Long.MIN_VALUE;
+			boolean overlaps = false;
+			for (Chunk chunk : chunks) {
+				long chunkStart = chunk.getChunkStart();
+				long chunkEnd = chunk.getChunkEnd();
+				if (overlaps(splitStart, splitEnd, chunkStart, chunkEnd)) {
+					long overlapStart = Math.max(splitStart, chunkStart);
+					long overlapEnd = Math.min(splitEnd, chunkEnd);
+					newStart = Math.min(newStart, overlapStart);
+					newEnd = Math.max(newEnd, overlapEnd);
+					overlaps = true;
+				}
+			}
+			if (overlaps) {
+				filteredSplits.add(new FileVirtualSplit(virtualSplit.getPath(), newStart, newEnd,
+						virtualSplit.getLocations()));
+			}
+		}
+		return filteredSplits;
+	}
+
+	private static boolean overlaps(long start, long end, long start2, long end2) {
```

Comment:
Should we use <code>CoordMath.overlaps(..)</code> instead here as well as below?